src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -2959,68 +2947,91 @@ CodeOffsets::Entries CompiledEntrySignature::c1_inline_ro_entry_type() const {
   }
 }
 
-void CompiledEntrySignature::compute_calling_conventions() {
-  // Get the (non-scalarized) signature and check for inline type arguments
+void CompiledEntrySignature::compute_calling_conventions(bool init) {
+  // Iterate over arguments and compute scalarized and non-scalarized signatures
+  bool has_scalarized = false;
   if (_method != NULL) {
+    InstanceKlass* holder = _method->method_holder();
+    int arg_num = 0;
     if (!_method->is_static()) {
-      if (_method->method_holder()->is_inline_klass() && InlineKlass::cast(_method->method_holder())->can_be_passed_as_fields()) {
+      if (holder->is_inline_klass() && InlineKlass::cast(holder)->can_be_passed_as_fields() &&
+          (init || _method->is_scalarized_arg(arg_num))) {
+        _sig_cc->appendAll(InlineKlass::cast(holder)->extended_sig());
+        has_scalarized = true;
         _has_inline_recv = true;
         _num_inline_args++;
+      } else {
+        SigEntry::add_entry(_sig_cc, T_OBJECT, holder->name());
       }
-      SigEntry::add_entry(_sig, T_OBJECT, _method->name());
+      SigEntry::add_entry(_sig, T_OBJECT, holder->name());
+      SigEntry::add_entry(_sig_cc_ro, T_OBJECT, holder->name());
+      arg_num++;
     }
     for (SignatureStream ss(_method->signature()); !ss.at_return_type(); ss.next()) {
       BasicType bt = ss.type();
-      if (bt == T_PRIMITIVE_OBJECT) {
-        if (ss.as_inline_klass(_method->method_holder())->can_be_passed_as_fields()) {
+      if (bt == T_OBJECT || bt == T_PRIMITIVE_OBJECT) {
+        InlineKlass* vk = ss.as_inline_klass(holder);
+        // TODO 8284443 Mismatch handling, we need to check parent method args (look at klassVtable::needs_new_vtable_entry)
+        if (vk != NULL && (bt == T_PRIMITIVE_OBJECT || holder->is_preload_class(vk->name())) &&
```

Comment:
Limiting argument scalarization to value classes present in the PreLoad attribute could lead to missed opportunities of scalarize if the class was loaded but not in the attribute (incomplete or missing PreLoad attributer case). The calling convention could instead look at which classes are loaded with no consideration to what triggered the loading. This would force overriding methods to look directly at the calling convention of the overridden method instead of looking at its context (lines 2861, 2882).
But this is a work that should be done when dealing with calling conventions mismatches (JDK-8284443).

Reply:
Yes, good point. As we discussed off-thread, I'll take care of this with JDK-8284443.